src/librustc_mir/interpret/memory.rs

```diff
@@ -772,6 +722,86 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
     }
 }
 
+#[doc(hidden)]
+/// There's no way to use this directly, it's just a helper struct for the `dump_alloc(s)` methods.
+pub struct DumpAllocs<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> {
+    mem: &'a Memory<'mir, 'tcx, M>,
+    allocs: Vec<AllocId>,
+}
+
+impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> std::fmt::Debug for DumpAllocs<'a, 'mir, 'tcx, M> {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // Cannot be a closure because it is generic in `Tag`, `Extra`.
+        fn write_allocation_track_relocs<'tcx, Tag: Copy + fmt::Debug, Extra>(
+            fmt: &mut std::fmt::Formatter<'_>,
+            tcx: TyCtxt<'tcx>,
+            allocs_to_print: &mut VecDeque<AllocId>,
+            alloc: &Allocation<Tag, Extra>,
+        ) -> std::fmt::Result {
+            for &(_, target_id) in alloc.relocations().values() {
+                allocs_to_print.push_back(target_id);
+            }
+            // This vec dance is necessary, because there is no trait
+            // that supports writing to files and to `std::fmt::Formatter` at the
+            // same time.
```

Comment:
Shouldn't using formatters also in `write_allocation` work? You can `write!(file, "{}", something_that_impls_Display)`.

Reply:
`write_allocation` has a generic parameter for the write-target, and you need some bound on that write target. Currently it has `io::Write`, which makes sense, since we use it for writing to files. But now that we want to write to `fmt::Formatter`, we'd need `fmt::Write`. Files (or stdout/stderr) for that matter, do not support `fmt::Write`, so we can't change it to `fmt::Write`. I could create a wrapper struct that impls `fmt::Write` for an inner `io::Write` object and wrap all file calls to `write_allocation`, then we can change the bounds on `write_allocation` to use `fmt::Write`.

Reply:
I think I was imagining an inverted pattern like here, where `write_allocation` *returns* `impl Display` which the caller then puts into the right stream.

Reply:
oh. Right. That can be done. :facepalm:

Reply:
Doing this now

Reply:
That worked like a charm! Thanks!